gh-112015: Implement `ctypes.memoryview_at()`

rianhunter opened the pull request:
ctypes currently has no way to easily create a buffer object from a pointer and a dynamic size. It is possible to create memoryview objects of array objects that refer to existing memory (e.g.
`memoryview((c_byte * size).from_address(address))`) but this is inefficient and the resulting memoryview object doesn't allow editing via python (invalid format errors are thrown, this is due to how ctypes objects implements the buffer protocol).

`ctypes.memoryview_at()` fills that gap in the API. This is similar to `ffi.buffer()` in the cffi module.

Fixes #112015 

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- readthedocs-preview cpython-previews start -->
----
:books: Documentation preview :books:: https://cpython-previews--112018.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

<!-- gh-issue-number: gh-112015 -->
* Issue: gh-112015
<!-- /gh-issue-number -->

rianhunter committed:
Implement `ctypes.buffer_at()`

ctypes currently has no way to easily create a buffer object
from a pointer and a dynamic size. It is possible to create
memoryview objects of array objects (e.g.
memoryview((c_ubyte * 10)())) but this is excessively slow when
implementing a callback function in Python that is passed a
dynamic void * and a size_t.

`ctypes.buffer_at()` fills that gap in the API. This is similar
to `ffi.buffer()` in the cffi module.

rianhunter committed:
Fix typo

serhiy-storchaka commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -5642,6 +5642,17 @@ wstring_at(const wchar_t *ptr, int size)
 }
 
+static PyObject *
+buffer_at(char *ptr, int size, int allow_write)
```
Why `int size` and not `Py_ssize_t size` or `size_t size`?

rianhunter replied:
For consistency with `string_at()`

serhiy-storchaka replied:
That code is perhaps outdated. And it is expected that `buffer_at()` will be used for larger data. There should not be a drawback of supporting larger buffers.

rianhunter replied:
You are right. Since I am not a frequent contributor, I erred on the side of consistency with the surrounding code.

rianhunter replied:
Okay I made this change but I used a `Py_ssize_t` instead because that is what is passed all the way through to `PyMemoryView_FromMemory()`.

serhiy-storchaka commented on Doc/library/ctypes.rst:
```diff
@@ -2109,6 +2109,19 @@ Utility functions
 
+.. function:: buffer_at(address, size, allow_write=False)
+
+   This function returns a memoryview object that references the
```
```suggestion
   This function returns a :class:`memoryview` object that references the
```

serhiy-storchaka commented on Doc/library/ctypes.rst:
```diff
@@ -2109,6 +2109,19 @@ Utility functions
+   This function returns a memoryview object that references the
+   memory starting at *address* up to (but not including) *address +
+   size*.  If *allow_write* is set to a truthy value then the
+   memoryview object is mutable.
```
```suggestion
   :class:`!memoryview` object is mutable.
```

serhiy-storchaka commented on Misc/NEWS.d/next/Library/2023-11-12-21-53-40.gh-issue-112015.2WPRxE.rst:
```diff
@@ -0,0 +1,3 @@
+ctypes.buffer_at() now exists to create a memoryview object that refers to
```
```suggestion
:func:`ctypes.buffer_at` now exists to create a :class:`memoryview` object that refers to
```

serhiy-storchaka commented on Misc/NEWS.d/next/Library/2023-11-12-21-53-40.gh-issue-112015.2WPRxE.rst:
```diff
@@ -0,0 +1,3 @@
+ctypes.buffer_at() now exists to create a memoryview object that refers to
+the supplied pointer and length. Works just like ctypes.string_at() except
```
```suggestion
the supplied pointer and length. Works just like :func:`ctypes.string_at` except
```

serhiy-storchaka commented on Doc/library/ctypes.rst:
```diff
@@ -2109,6 +2109,19 @@ Utility functions
+   of not making a copy of the specified memory.
+
+   .. audit-event:: ctypes.buffer_at address,size,allow_write ctypes.buffer_at
+
```
```suggestion

   .. versionadded:: 3.13

```

serhiy-storchaka reviewed (commented):
Please add tests and What's New entry.

And I think that it is worth to support large buffers.

rianhunter committed:
Apply suggestions from code review

Documentation improvements

Co-authored-by: Serhiy Storchaka <storchaka@gmail.com>

rianhunter committed:
Make size argument a Py_ssize_t, per @serhiy-storchaka suggestion

rianhunter committed:
Add what's new entry

rianhunter committed:
Rename buffer_at to memoryview_at

It's a more descriptive and precise name.

rianhunter committed:
Make mutable objects the default

It's more common that the user will want a memoryview object that can
be mutated. An immutable object is more rare, so make the default case
return a mutable object.

rianhunter committed:
Add test for ctypes.memoryview_at

encukou committed:
Merge in the main branch; move What's New entry

encukou commented on Lib/test/test_ctypes/test_memfunctions.py:
```diff
@@ -77,6 +78,25 @@ def test_wstring_at(self):
+        # by C APIs. It's an efficient way to get a buffer
+        # representing a dynamically-sized memory region without having
+        # to create an array type first.
+        v = memoryview_at(foreign_ptr, foreign_ptr_size)
```
Shouldn't the first argument be an *address*, rather than a ctypes pointer?

rianhunter replied:
It can be either just like with `string_at()`. It's intended to be used with `c_void_p` objects.

encukou replied:
Could you add it to the test and docs?

I know the existing tests and docs are lacking, but let's do better today.

rianhunter replied:
I'm sorry but I no longer have time to work on this patch. Sorry for wasting your time

encukou committed:
TMP

encukou committed:
Merge in the main branch

encukou committed:
Revert to calling through `ctypes` to get `c_void_p` conversion semantics. Improve tests and docs.

encukou committed:
Test read-only memoryview

encukou committed:
Use c_ssize_t for the size

encukou committed:
Test size overflow

encukou committed:
Doc fixups. Don't imply that *readonly* makes the memory immutable.

encukou committed:
Fixups

encukou committed:
Merge branch 'main' into ctypes-buffer-at

picnixz commented on Doc/whatsnew/3.14.rst:
```diff
@@ -326,6 +326,13 @@ ctypes
+  :class:`memoryview` object that refers to the supplied pointer and
+  length. Works just like :func:`ctypes.string_at` except it avoids a
+  buffer copy. Useful when implementing callback functions in Python
+  that are passed dynamically-sized buffers.
```
```suggestion
  length. This works like :func:`ctypes.string_at` except it avoids a
  buffer copy, and is typically useful when implementing pure Python
  callback functions that are passed dynamically-sized buffers.
```

picnixz commented on Doc/whatsnew/3.14.rst:
```diff
@@ -326,6 +326,13 @@ ctypes
+  buffer copy. Useful when implementing callback functions in Python
+  that are passed dynamically-sized buffers.
+  (Contributed by Rian Hunter in :gh:`112018`.)
+
```
```suggestion


```

picnixz commented on Lib/ctypes/__init__.py:
```diff
@@ -524,6 +524,7 @@ def WinError(code=None, descr=None):
 # functions
 
 from _ctypes import _memmove_addr, _memset_addr, _string_at_addr, _cast_addr
+from _ctypes import _memoryview_at_addr
```
Why not merge the above imports? is it to avoid a line continuation?

encukou replied:
Yes.

picnixz commented on Lib/ctypes/__init__.py:
```diff
@@ -549,6 +550,14 @@ def string_at(ptr, size=-1):
+def memoryview_at(ptr, size, readonly=False):
+    """memoryview_at(ptr[, size]) -> memoryview
+
+    Return a memoryview representing the memory at addr."""
```
```suggestion
    Return a memoryview representing the memory at void *ptr."""
```

picnixz commented on Lib/ctypes/__init__.py:
```diff
@@ -549,6 +550,14 @@ def string_at(ptr, size=-1):
+_memoryview_at = PYFUNCTYPE(
+    py_object, c_void_p, c_ssize_t, c_int)(_memoryview_at_addr)
+def memoryview_at(ptr, size, readonly=False):
+    """memoryview_at(ptr[, size]) -> memoryview
```
```suggestion
    """memoryview_at(ptr, size[, readonly]) -> memoryview
```

picnixz commented on Misc/NEWS.d/next/Library/2023-11-12-21-53-40.gh-issue-112015.2WPRxE.rst:
```diff
@@ -0,0 +1,5 @@
+:func:`ctypes.memoryview_at` now exists to create a
```
Don't forget the sync the NEWS and the What's New if you accept my suggestion.

picnixz commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -5741,6 +5741,23 @@ wstring_at(const wchar_t *ptr, int size)
+static PyObject *
+memoryview_at(void *ptr, Py_ssize_t size, int readonly)
+{
+    Py_ssize_t ssize = size;
```
Why this additional conversion? we are given a `ssize_t` as input so it wouldn't change anything here, unless you want that variable only for semantic purposes.

encukou replied:
Thanks! Yes, I left that in from `wstring_at`

encukou commented on Misc/NEWS.d/next/Library/2023-11-12-21-53-40.gh-issue-112015.2WPRxE.rst:
```diff
@@ -0,0 +1,5 @@
+:class:`memoryview` object that refers to the supplied pointer and
+length. Works just like :func:`ctypes.string_at` except it avoids a
+buffer copy. Useful when implementing callback functions in Python
+that are passed dynamically-sized buffers.
```
```suggestion
length. This works like :func:`ctypes.string_at` except it avoids a
buffer copy, and is typically useful when implementing pure Python
callback functions that are passed dynamically-sized buffers.
```

encukou committed:
Avoid extra variable

encukou committed:
Apply suggestions from code review

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

encukou committed:
Merge in the main branch

ZeroIntensity commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -5747,6 +5747,22 @@ wstring_at(const wchar_t *ptr, int size)
+memoryview_at(void *ptr, Py_ssize_t size, int readonly)
+{
+    if (PySys_Audit("ctypes.memoryview_at", "nni",
+                    (Py_ssize_t)ptr, size, readonly) < 0) {
```
Is there a possibility of data loss with this cast if `sizeof(Py_ssize_t) != sizeof(intptr_t)`?

encukou replied:
For the audit? Yes. `n` is the closest `PySys_Audit`/`PyBuildValue` have to `intptr_t`, though. 
There's a bunch of these across ctypes. If we want to change it, should go in a separate PR.

ZeroIntensity commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -5747,6 +5747,22 @@ wstring_at(const wchar_t *ptr, int size)
+        return NULL;
+    }
+    if (size < 0) {
+        PyErr_Format(PyExc_ValueError,
```
Actually, do we want to check this before the audit?

encukou replied:
IMO, people who look at audit reports won't mind having logs of unsuccessful attempts :)

ZeroIntensity commented on Lib/test/test_ctypes/test_memfunctions.py:
```diff
@@ -77,6 +79,64 @@ def test_wstring_at(self):
+    def test_memoryview_at_readonly(self):
+        b = (c_byte * 10)()
+
+        foreign_ptr = cast(b, c_void_p)
```
Doesn't this just get overwritten by the loop?

ZeroIntensity reviewed (commented):
Mostly LGTM

encukou committed:
Remove unneeded line

encukou committed:
Merge in the main branch

ZeroIntensity commented on Lib/test/test_ctypes/test_memfunctions.py:
```diff
@@ -77,6 +79,63 @@ def test_wstring_at(self):
+    def test_memoryview_at(self):
+        b = (c_byte * 10)()
+
+        foreign_ptr = cast(b, c_void_p)
```
Same here, then LGTM.

encukou committed:
Merge in the main branch

encukou committed:
Remove unneeded line

ZeroIntensity reviewed (approved):
I'm happy with this!
